### PR TITLE
DOC: Add instructions for building with ASAN

### PR DESCRIPTION
**Pull request by david-cortes-intel**
## Description

This PR adds instructions for building sklearnex with ASAN (assuming a oneDAL build with it too) and running python with it.

---

PR should start as a draft, then move to ready for review state after CI is passed and all applicable checkboxes are closed.
This approach ensures that reviewers don't spend extra time asking for regular requirements.

You can remove a checkbox as not applicable only if it doesn't relate to this PR in any way.
For example, PR with docs update doesn't require checkboxes for performance while PR with any change in actual code should have checkboxes and justify how this code change is expected to affect performance (or justification should be self-evident).

Checklist to comply with **before moving PR from draft**:

**PR completeness and readability**

- [x] I have reviewed my changes thoroughly before submitting this pull request.
- [x] I have updated the documentation to reflect the changes or created a separate PR with update and provided its number in the description, if necessary.
- [x] Git commit message contains an appropriate signed-off-by string _(see [CONTRIBUTING.md](https://github.com/uxlfoundation/scikit-learn-intelex/blob/main/CONTRIBUTING.md#pull-requests) for details)_.
- [x] I have added a respective label(s) to PR if I have a permission for that.
- [x] I have resolved any merge conflicts that might occur with the base branch.

**Testing**

- [x] I have run it locally and tested the changes extensively.
- [x] All CI jobs are green or I have provided justification why they aren't.

### CODE REVIEW

File: INSTALL.md [hunk 1]
```diff
@@ -226,6 +226,45 @@ python setup.py build --abs-rpath
...
+
+Then, the Python memory allocator can be set to `malloc` like this:
+```shell
+export PYTHONMALLOC=malloc
+```
+
+Putting it all together, the earlier examples building the library in-place and executing a python file with it become as follows:
+```shell
+source <path to ASAN-enabled oneDAL env.sh>
+CC="icx -fsanitize=address -g" CXX="icpx -fsanitize=address -g" LDFLAGS="-static-libasan" python setup.py build_ext --inplace --force --abs-rpath
```
Comment (napetrov) [inline INSTALL.md:261]: should it be ICX?

Comment (david-cortes-intel) [reply INSTALL.md:261]: If you meant about the `CXX` variable, then: no, on linux the C++ compiler is invoked through `icpx`.

Comment (napetrov) [reply INSTALL.md:261]: i mean that we would need  icx only for gpu portion of scikit but CPU side of the things should be gcc

Comment (david-cortes-intel) [reply INSTALL.md:261]: Yes, but then you'd be mixing a library compiled with ICX (oneDAL) expecting the clang runtime for ASAN, and a library compiled with GCC expecting the gnu runtime for ASAN.

It does work with swapped runtimes as far as I can tell, but I think the clang runtime has additional things that the gnu one doesn't, so it could potentially lead to incompatibility problems depending on which functionalities/symbols/etc. from ASAN get used.

File: INSTALL.md [hunk 2]
```diff
@@ -226,6 +226,45 @@ python setup.py build --abs-rpath
...
+See the instructions on the oneDAL repository for building the library from source with ASAN enabled:
+https://github.com/uxlfoundation/oneDAL/blob/main/INSTALL.md
+
+When building `scikit-learn-intelex`, the system's default compiler is used unless specified otherwise through variables such as `$CXX`. In order to avoid issues with incompatible runtimes of ASAN, one might want to change the compiler to ICX if oneDAL was built with ICX (the default for it).
+
+The compiler and flags to build with both ASAN and debug symbols can be controlled through environment variables - **assuming a Linux system** (ASAN on Windows has not been tested):
+```shell
+export CC="icx -fsanitize=address -g"
+export CXX="icpx -fsanitize=address -g"
+export LDFLAGS="-static-libasan"
```
Comment (yuejiaointel) [inline INSTALL.md:245]: One naive question: why we need to link ASAN statically?

Comment (david-cortes-intel) [reply INSTALL.md:245]: I don't know, but it doesn't work otherwise.

Comment (david-cortes-intel) [reply INSTALL.md:245]: Actually turns out it's not needed when using ICX, will remove it.

File: INSTALL.md [hunk 3]
```diff
@@ -226,6 +226,44 @@ python setup.py build --abs-rpath
 
 **Note:** when building `scikit-learn-intelex` from source with this option, it will use the oneDAL library with which it was compiled. oneDAL has dependencies on other libraries such as TBB, which is also distributed as a python package through `pip` and as a `conda` package. By default, a conda environment will first try to load TBB from its own packages if it is installed in the environment, which might cause issues if oneDAL was compiled with a system TBB instead of a conda one. In such cases, it is advised to either uninstall TBB from pip/conda (it will be loaded from the oneDAL library which links to it), or modify the order of search paths in environment variables like `${LD_LIBRARY_PATH}`.
 
+### Building with ASAN
+
+In order to use the address sanitizer (ASAN) together with `scikit-learn-intelex`, it's necessary to:
```
Comment (ethanglaser) [inline INSTALL.md:231]:
```suggestion
In order to use AddressSanitizer (ASAN) together with `scikit-learn-intelex`, it's necessary to:
```
Based on how they refer to it in https://github.com/google/sanitizers/wiki/addresssanitizer

**Review by ethanglaser — COMMENTED** [review]
This looks comprehensive, thanks for creating. General question (since I am not super familiar) - is this something geared more towards users or contributors? INSTALL.md doesn't seem like the right location, maybe CONTRIBUTING.md or within docs/ ? Would welcome other opinions on this

**Comment by david-cortes-intel** [comment]
> This looks comprehensive, thanks for creating. General question (since I am not super familiar) - is this something geared more towards users or contributors? INSTALL.md doesn't seem like the right location, maybe CONTRIBUTING.md or within docs/ ? Would welcome other opinions on this

This is more geared towards contributors.

It is indeed the case that our docs have things mixed up throughout different places and would benefit from reordering things and deduplicating sections, but currently this is the only place having instructions for executing the setup.py file, and the instructions here are easier to read when put next to the instructions for building without ASAN.

**Comment by ethanglaser** [comment]
> > This looks comprehensive, thanks for creating. General question (since I am not super familiar) - is this something geared more towards users or contributors? INSTALL.md doesn't seem like the right location, maybe CONTRIBUTING.md or within docs/ ? Would welcome other opinions on this
> 
> This is more geared towards contributors.
> 
> It is indeed the case that our docs have things mixed up throughout different places and would benefit from reordering things and deduplicating sections, but currently this is the only place having instructions for executing the setup.py file, and the instructions here are easier to read when put next to the instructions for building without ASAN.

Makes sense, overall I think the content is ready for merge but would be interested if anyone else has thoughts on location of this info.

File: INSTALL.md [hunk 4]
```diff
@@ -214,6 +214,44 @@ python setup.py build --abs-rpath
...
+
+When building `scikit-learn-intelex`, the system's default compiler is used unless specified otherwise through variables such as `$CXX`. In order to avoid issues with incompatible runtimes of ASAN, one might want to change the compiler to ICX if oneDAL was built with ICX (the default for it).
+
+The compiler and flags to build with both ASAN and debug symbols can be controlled through environment variables - **assuming a Linux system** (ASAN on Windows has not been tested):
+```shell
+export CC="icx -fsanitize=address -g"
+export CXX="icpx -fsanitize=address -g"
+```
+
+The ASAN runtime used by ICX is the same as the one by CLANG. It's possible to preload the ASAN runtime for GNU if that's the system's default through e.g. `LD_PRELOAD=libasan.so` or similar, but to get the same ASAN runtime as for oneDAL, one might need to specifically pass the paths from CLANG if that's not the system's default compiler:
```
Comment (icfaust) [inline INSTALL.md]:
```suggestion
The ASAN runtime used by ICX is the same as the one by Clang. It's possible to preload the ASAN runtime for GNU if that's the system's default through e.g. `LD_PRELOAD=libasan.so` or similar, but to get the same ASAN runtime as for oneDAL, one might need to specifically pass the paths from Clang if that's not the system's default compiler:
```

Nitpick about how Clang is written (https://clang.llvm.org/) following how they refer to it as 'Clang'

File: INSTALL.md [hunk 5]
```diff
@@ -214,6 +214,44 @@ python setup.py build --abs-rpath
...
+```
+
+Then, the Python memory allocator can be set to `malloc` like this:
+```shell
+export PYTHONMALLOC=malloc
+```
+
+Putting it all together, the earlier examples building the library in-place and executing a python file with it become as follows:
+```shell
+source <path to ASAN-enabled oneDAL env.sh>
```
Comment (icfaust) [inline INSTALL.md]:
```suggestion
source <path to ASan-enabled oneDAL env.sh>
```
Similar for ASan : https://github.com/google/sanitizers/wiki/addresssanitizer

File: INSTALL.md [hunk 6]
```diff
@@ -214,6 +214,44 @@ python setup.py build --abs-rpath
...
+CC="icx -fsanitize=address -g" CXX="icpx -fsanitize=address -g" python setup.py build_ext --inplace --force --abs-rpath
+CC="icx -fsanitize=address -g" CXX="icpx -fsanitize=address -g" python setup.py build --abs-rpath
+LD_PRELOAD="$(clang -print-file-name=libclang_rt.asan-x86_64.so)" PYTHONMALLOC=malloc PYTHONPATH=$(pwd) python <python file.py>
+```
+
+_Be aware that ASAN is known to generate many false-positive reports of memory leaks when used with oneDAL, NumPy, and SciPy._
```
Comment (icfaust) [inline INSTALL.md]:
```suggestion
_Be aware that ASan is known to generate many false-positive reports of memory leaks when used with oneDAL, NumPy, and SciPy._
```

File: INSTALL.md [hunk 7]
```diff
@@ -214,6 +214,44 @@ python setup.py build --abs-rpath
 
 **Note:** when building `scikit-learn-intelex` from source with this option, it will use the oneDAL library with which it was compiled. oneDAL has dependencies on other libraries such as TBB, which is also distributed as a python package through `pip` and as a `conda` package. By default, a conda environment will first try to load TBB from its own packages if it is installed in the environment, which might cause issues if oneDAL was compiled with a system TBB instead of a conda one. In such cases, it is advised to either uninstall TBB from pip/conda (it will be loaded from the oneDAL library which links to it), or modify the order of search paths in environment variables like `${LD_LIBRARY_PATH}`.
 
+### Building with ASAN
+
+In order to use AddressSanitizer (ASAN) together with `scikit-learn-intelex`, it's necessary to:
+* Build both oneDAL and scikit-learn-intelex with ASAN (otherwise error traces will not be very informative).
+* Preload the ASAN runtime when executing the Python process that imports `scikit-learn-intelex`.
```
Comment (icfaust) [inline INSTALL.md]:
```suggestion
In order to use AddressSanitizer (ASan) together with `scikit-learn-intelex`, it's necessary to:
* Build both oneDAL and scikit-learn-intelex with ASan (otherwise error traces will not be very informative).
* Preload the ASan runtime when executing the Python process that imports `scikit-learn-intelex`.
```

File: INSTALL.md [hunk 8]
```diff
@@ -214,6 +214,44 @@ python setup.py build --abs-rpath
...
+* Optionally, configure Python to use `malloc` as default allocator to reduce the number of false-positive leak reports.
+
+See the instructions on the oneDAL repository for building the library from source with ASAN enabled:
+https://github.com/uxlfoundation/oneDAL/blob/main/INSTALL.md
+
+When building `scikit-learn-intelex`, the system's default compiler is used unless specified otherwise through variables such as `$CXX`. In order to avoid issues with incompatible runtimes of ASAN, one might want to change the compiler to ICX if oneDAL was built with ICX (the default for it).
+
+The compiler and flags to build with both ASAN and debug symbols can be controlled through environment variables - **assuming a Linux system** (ASAN on Windows has not been tested):
```
Comment (icfaust) [inline INSTALL.md]:
```suggestion
When building `scikit-learn-intelex`, the system's default compiler is used unless specified otherwise through variables such as `$CXX`. In order to avoid issues with incompatible runtimes of ASan, one might want to change the compiler to ICX if oneDAL was built with ICX (the default for it).

The compiler and flags to build with both ASan and debug symbols can be controlled through environment variables - **assuming a Linux system** (ASan on Windows has not been tested):
```

File: INSTALL.md [hunk 9]
```diff
@@ -214,6 +214,44 @@ python setup.py build --abs-rpath
...
+
+When building `scikit-learn-intelex`, the system's default compiler is used unless specified otherwise through variables such as `$CXX`. In order to avoid issues with incompatible runtimes of ASan, one might want to change the compiler to ICX if oneDAL was built with ICX (the default for it).
+
+The compiler and flags to build with both ASan and debug symbols can be controlled through environment variables - **assuming a Linux system** (ASan on Windows has not been tested):
+```shell
+export CC="icx -fsanitize=address -g"
+export CXX="icpx -fsanitize=address -g"
+```
+
+The ASAN runtime used by ICX is the same as the one by Clang. It's possible to preload the ASAN runtime for GNU if that's the system's default through e.g. `LD_PRELOAD=libasan.so` or similar, but to get the same ASAN runtime as for oneDAL, one might need to specifically pass the paths from Clang if that's not the system's default compiler:
```
Comment (icfaust) [inline INSTALL.md]:
```suggestion
The ASAN runtime used by ICX is the same as the one by Clang. It's possible to preload the ASan runtime for GNU if that's the system's default through e.g. `LD_PRELOAD=libasan.so` or similar, but to get the same ASan runtime as for oneDAL, one might need to specifically pass the paths from Clang if that's not the system's default compiler:
```
